File: packages/imanikurd-prayer/src/prayer.ts
import { KURDISTAN_CITIES, getCityByCoordinates, getCityDisplayName } from "./cities.js";
import { loadData, loadDataAsync } from "./loadData.js";
import type {
  DayPrayerRecord,
  NextPrayer,
  PrayerTimings,
  PrayerTimesData,
} from "./types.js";

let prayerCache: PrayerTimesData | null = null;

function getPrayerData(): PrayerTimesData {
  if (!prayerCache) {
    prayerCache = loadData<PrayerTimesData>("prayer_times.json");
  }
  return prayerCache;
}

async function getPrayerDataAsync(): Promise<PrayerTimesData> {
  if (!prayerCache) {
    prayerCache = await loadDataAsync<PrayerTimesData>("prayer_times.json");
  }
  return prayerCache;
}

function formatDateKey(date: Date): string {
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const day = String(date.getDate()).padStart(2, "0");
  return `${month}-${day}`;
}

function toTimings(record: DayPrayerRecord): PrayerTimings {
  return {
    Fajr: record.fajr,
    Sunrise: record.sunrise,
    Dhuhr: record.dhuhr,
    Asr: record.asr,
    Maghrib: record.maghrib,
    Isha: record.isha,
  };
}

function parseTimeToMinutes(timeStr: string): number {
  if (!timeStr) return 0;
  const cleaned = timeStr.split(" ")[0].trim();
  const [h, m] = cleaned.split(":").map(Number);
  if (isNaN(h) || isNaN(m)) return 0;
  return h * 60 + m;
}

/** Get all available city keys for prayer times */
export function getPrayerCities(): string[] {
  return Object.keys(getPrayerData()).sort();
}

/** Get all available city keys for prayer times asynchronously */
export async function getPrayerCitiesAsync(): Promise<string[]> {
  const data = await getPrayerDataAsync();
  return Object.keys(data).sort();
}

/** Get prayer times for a city on a specific date (defaults to today) */
export function getPrayerTimes(
  cityKey: string,
  date: Date = new Date()
): PrayerTimings | null {
  const data = getPrayerData();
  const cityRecords = data[cityKey] || data["Hawler"];
  if (!cityRecords) return null;

  const dateKey = formatDateKey(date);
  const record = cityRecords.find((r) => r.date === dateKey);
  return record ? toTimings(record) : null;
}

/** Get prayer times for a city on a specific date asynchronously */
export async function getPrayerTimesAsync(
  cityKey: string,
  date: Date = new Date()
): Promise<PrayerTimings | null> {
  const data = await getPrayerDataAsync();
  const cityRecords = data[cityKey] || data["Hawler"];
  if (!cityRecords) return null;

  const dateKey = formatDateKey(date);
  const record = cityRecords.find((r) => r.date === dateKey);
  return record ? toTimings(record) : null;
}

/** Get full daily record including date string */
export function getDailyPrayerRecord(
  cityKey: string,
  date: Date = new Date()
): (DayPrayerRecord & { timings: PrayerTimings }) | null {
  const data = getPrayerData();
  const cityRecords = data[cityKey] || data["Hawler"];
  if (!cityRecords) return null;

  const dateKey = formatDateKey(date);
  const record = cityRecords.find((r) => r.date === dateKey);
  if (!record) return null;

  return { ...record, timings: toTimings(record) };
}

/** Get full daily record including date string asynchronously */
export async function getDailyPrayerRecordAsync(
  cityKey: string,
  date: Date = new Date()
): Promise<(DayPrayerRecord & { timings: PrayerTimings }) | null> {
  const data = await getPrayerDataAsync();
  const cityRecords = data[cityKey] || data["Hawler"];
  if (!cityRecords) return null;

  const dateKey = formatDateKey(date);
  const record = cityRecords.find((r) => r.date === dateKey);
  if (!record) return null;

  return { ...record, timings: toTimings(record) };
}

/** Get all prayer times for a city in a given month */
export function getMonthlyPrayerTimes(
  cityKey: string,
  month: number,
  year: number = new Date().getFullYear()
) {
  const data = getPrayerData();
  const cityRecords = data[cityKey] || data["Hawler"] || [];

  return cityRecords
    .filter((record) => {
      const [mm] = record.date.split("-").map(Number);
      return mm === month;
    })
    .map((record) => {
      const [mm, dd] = record.date.split("-").map(Number);
      return {
        date: new Date(year, mm - 1, dd),
        day: dd,
        month: mm,
        timings: toTimings(record),
      };
    })
    .sort((a, b) => a.day - b.day);
}

/** Get all prayer times for a city in a given month asynchronously */
export async function getMonthlyPrayerTimesAsync(
  cityKey: string,
  month: number,
  year: number = new Date().getFullYear()
) {
  const data = await getPrayerDataAsync();
  const cityRecords = data[cityKey] || data["Hawler"] || [];

  return cityRecords
    .filter((record) => {
      const [mm] = record.date.split("-").map(Number);
      return mm === month;
    })
    .map((record) => {
      const [mm, dd] = record.date.split("-").map(Number);
      return {
        date: new Date(year, mm - 1, dd),
        day: dd,
        month: mm,
        timings: toTimings(record),
      };
    })
    .sort((a, b) => a.day - b.day);
}

/** Detect nearest Kurdistan city from coordinates */
export function getCityFromCoordinates(lat: number, lng: number): string {
  return getCityByCoordinates(lat, lng).key;
}

/** Get Kurdish display name for a city key */
export function getCityName(cityKey: string): string {
  return getCityDisplayName(cityKey);
}

/** Get the next upcoming prayer for a city */
export function getNextPrayer(
  cityKey: string,
  now: Date = new Date()
): NextPrayer | null {
  const timings = getPrayerTimes(cityKey, now);
  if (!timings) return null;

  const prayers = [
    { name: "Fajr", time: timings.Fajr },
    { name: "Sunrise", time: timings.Sunrise },
    { name: "Dhuhr", time: timings.Dhuhr },
    { name: "Asr", time: timings.Asr },
    { name: "Maghrib", time: timings.Maghrib },
    { name: "Isha", time: timings.Isha },
  ];

  const currentMins = now.getHours() * 60 + now.getMinutes();

  for (const prayer of prayers) {
    const prayerMins = parseTimeToMinutes(prayer.time);
    if (prayerMins > currentMins) {
      const diff = prayerMins - currentMins;
      return {
        name: prayer.name,
        time: prayer.time,
        hoursUntil: Math.floor(diff / 60),
        minutesUntil: diff % 60,
      };
    }
  }

  const fajrMins = parseTimeToMinutes(prayers[0].time);
  const diff = 24 * 60 - currentMins + fajrMins;
  return {
    name: prayers[0].name,
    time: prayers[0].time,
    hoursUntil: Math.floor(diff / 60),
    minutesUntil: diff % 60,
  };
}

/** Get the next upcoming prayer for a city asynchronously */
export async function getNextPrayerAsync(
  cityKey: string,
  now: Date = new Date()
): Promise<NextPrayer | null> {
  const timings = await getPrayerTimesAsync(cityKey, now);
  if (!timings) return null;

  const prayers = [
    { name: "Fajr", time: timings.Fajr },
    { name: "Sunrise", time: timings.Sunrise },
    { name: "Dhuhr", time: timings.Dhuhr },
    { name: "Asr", time: timings.Asr },
    { name: "Maghrib", time: timings.Maghrib },
    { name: "Isha", time: timings.Isha },
  ];

  const currentMins = now.getHours() * 60 + now.getMinutes();

  for (const prayer of prayers) {
    const prayerMins = parseTimeToMinutes(prayer.time);
    if (prayerMins > currentMins) {
      const diff = prayerMins - currentMins;
      return {
        name: prayer.name,
        time: prayer.time,
        hoursUntil: Math.floor(diff / 60),
        minutesUntil: diff % 60,
      };
    }
  }

  const fajrMins = parseTimeToMinutes(prayers[0].time);
  const diff = 24 * 60 - currentMins + fajrMins;
  return {
    name: prayers[0].name,
    time: prayers[0].time,
    hoursUntil: Math.floor(diff / 60),
    minutesUntil: diff % 60,
  };
}

/** Get prayer times using geolocation (auto-detect city) */
export function getPrayerTimesByLocation(
  lat: number,
  lng: number,
  date: Date = new Date()
): PrayerTimings | null {
  const cityKey = getCityFromCoordinates(lat, lng);
  return getPrayerTimes(cityKey, date);
}

/** Get prayer times using geolocation (auto-detect city) asynchronously */
export async function getPrayerTimesByLocationAsync(
  lat: number,
  lng: number,
  date: Date = new Date()
): Promise<PrayerTimings | null> {
  const cityKey = getCityFromCoordinates(lat, lng);
  return getPrayerTimesAsync(cityKey, date);
}

export { KURDISTAN_CITIES };

export type { DayPrayerRecord, NextPrayer, PrayerTimings, PrayerTimesData };

File: packages/imanikurd_flutter/lib/src/prayer.dart
import 'cities.dart';
import 'load_data.dart';
import 'models.dart';

typedef PrayerTimesData = Map<String, List<DayPrayerRecord>>;

PrayerTimesData? _prayerCache;

Future<PrayerTimesData> _getPrayerData() async {
  if (_prayerCache != null) return _prayerCache!;
  final raw = await loadDataAsync<Map<String, dynamic>>('prayer_times.json');
  _prayerCache = raw.map(
    (city, records) => MapEntry(
      city,
      (records as List<dynamic>)
          .map((e) => DayPrayerRecord.fromJson(e as Map<String, dynamic>))
          .toList(),
    ),
  );
  return _prayerCache!;
}

String _formatDateKey(DateTime date) {
  final month = date.month.toString().padLeft(2, '0');
  final day = date.day.toString().padLeft(2, '0');
  return '$month-$day';
}

int _parseTimeToMinutes(String timeStr) {
  if (timeStr.isEmpty) return 0;
  final cleaned = timeStr.split(' ').first.trim();
  final parts = cleaned.split(':');
  if (parts.length < 2) return 0;
  final h = int.tryParse(parts[0]);
  final m = int.tryParse(parts[1]);
  if (h == null || m == null) return 0;
  return h * 60 + m;
}

Future<List<String>> getPrayerCities() async {
  final data = await _getPrayerData();
  final keys = data.keys.toList()..sort();
  return keys;
}

Future<PrayerTimings?> getPrayerTimes(String cityKey, [DateTime? date]) async {
  final data = await _getPrayerData();
  final cityRecords = data[cityKey] ?? data['Hawler'];
  if (cityRecords == null) return null;

  final dateKey = _formatDateKey(date ?? DateTime.now());
  for (final record in cityRecords) {
    if (record.date == dateKey) return record.toTimings();
  }
  return null;
}

Future<DailyPrayerRecordWithTimings?> getDailyPrayerRecord(String cityKey, [DateTime? date]) async {
  final data = await _getPrayerData();
  final cityRecords = data[cityKey] ?? data['Hawler'];
  if (cityRecords == null) return null;

  final dateKey = _formatDateKey(date ?? DateTime.now());
  for (final record in cityRecords) {
    if (record.date == dateKey) {
      return DailyPrayerRecordWithTimings(
        date: record.date,
        fajr: record.fajr,
        sunrise: record.sunrise,
        dhuhr: record.dhuhr,
        asr: record.asr,
        maghrib: record.maghrib,
        isha: record.isha,
        timings: record.toTimings(),
      );
    }
  }
  return null;
}

Future<List<MonthlyPrayerDay>> getMonthlyPrayerTimes(
  String cityKey,
  int month, [
  int? year,
]) async {
  final data = await _getPrayerData();
  final cityRecords = data[cityKey] ?? data['Hawler'] ?? [];
  final targetYear = year ?? DateTime.now().year;

  final result = <MonthlyPrayerDay>[];
  for (final record in cityRecords) {
    final parts = record.date.split('-').map(int.parse).toList();
    if (parts.length < 2) continue;
    final mm = parts[0];
    final dd = parts[1];
    if (mm == month) {
      result.add(
        MonthlyPrayerDay(
          date: DateTime(targetYear, mm, dd),
          day: dd,
          month: mm,
          timings: record.toTimings(),
        ),
      );
    }
  }
  result.sort((a, b) => a.day.compareTo(b.day));
  return result;
}

String getCityFromCoordinates(double lat, double lng) => getCityByCoordinates(lat, lng).key;

String getCityName(String cityKey) => getCityDisplayName(cityKey);

Future<NextPrayer?> getNextPrayer(String cityKey, [DateTime? now]) async {
  final current = now ?? DateTime.now();
  final timings = await getPrayerTimes(cityKey, current);
  if (timings == null) return null;

  final prayers = [
    ('Fajr', timings.fajr),
    ('Sunrise', timings.sunrise),
    ('Dhuhr', timings.dhuhr),
    ('Asr', timings.asr),
    ('Maghrib', timings.maghrib),
    ('Isha', timings.isha),
  ];

  final currentMins = current.hour * 60 + current.minute;

  for (final prayer in prayers) {
    final prayerMins = _parseTimeToMinutes(prayer.$2);
    if (prayerMins > currentMins) {
      final diff = prayerMins - currentMins;
      return NextPrayer(
        name: prayer.$1,
        time: prayer.$2,
        hoursUntil: diff ~/ 60,
        minutesUntil: diff % 60,
      );
    }
  }

  final fajrMins = _parseTimeToMinutes(prayers.first.$2);
  final diff = 24 * 60 - currentMins + fajrMins;
  return NextPrayer(
    name: prayers.first.$1,
    time: prayers.first.$2,
    hoursUntil: diff ~/ 60,
    minutesUntil: diff % 60,
  );
}

Future<PrayerTimings?> getPrayerTimesByLocation(
  double lat,
  double lng, [
  DateTime? date,
]) {
  final cityKey = getCityFromCoordinates(lat, lng);
  return getPrayerTimes(cityKey, date);
}
